### PR TITLE
PP-4140 Add first 6 digits as search param

### PR DESCRIPTION
**Pull request by simad**
## WHAT
We can now search by first 6 digits of card number.
- Added "type" object to strongly type LastDigits and FirstDigits instead of Strings. These classes should be moved to pay java commons, I've created a ticket for that.
- Added validation of length / numeric string inside the object themselves. I think it simplifies things, because we either have valid objects or nothing. This is open to discussion, of course.

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/model/LastDigitsCardNumber.java [hunk 1]
```diff
@@ -1,16 +1,30 @@
...
-        this.lastDigitsCardNumber = Objects.requireNonNull(lastDigitsCardNumber);
+        this.lastDigitsCardNumber = lastDigitsCardNumber;
     }
 
+    private static boolean isValid(String lastDigitsCardNumber) {
+        return lastDigitsCardNumber != null && lastDigitsCardNumber.length() == 4 && StringUtils.isNumeric(lastDigitsCardNumber);
+    }
+    
     public static LastDigitsCardNumber of(String lastDigitsCardNumber) {
+        if (!(isValid(lastDigitsCardNumber))) {
```
Comment (DanailMinchev) [inline src/main/java/uk/gov/pay/connector/model/LastDigitsCardNumber.java:18]: can be `if (!isValid(lastDigitsCardNumber)` without extra `()` but I don't mind

File: src/test/java/uk/gov/pay/connector/dao/ChargeSearchParamsTest.java [hunk 2]
```diff
@@ -68,7 +75,7 @@ public void buildQueryParams_chargeSearch_withAllParameters() throws Exception {
     }
 
     @Test
-    public void buildQueryParamsWithPiiRedaction_chargeSearch_withAllParameters() throws Exception {
```
Comment (DanailMinchev) [inline src/test/java/uk/gov/pay/connector/dao/ChargeSearchParamsTest.java:71]: 👍

File: src/test/java/uk/gov/pay/connector/model/FirstDigitsCardNumberTest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        Assert.assertThat(FirstDigitsCardNumber.ofNullable("123456").toString(), is("123456"));
+    }
+
+    @Test
+    public void shouldReturnNullIfStringIsNull() {
+        Assert.assertThat(FirstDigitsCardNumber.ofNullable(null), is(nullValue()));
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void shouldThrowIfNonNumericFirstSixDigitsOfCard() {
```
Comment (DanailMinchev) [inline src/test/java/uk/gov/pay/connector/model/FirstDigitsCardNumberTest.java:24]: Great test coverage! I was wondering should we have a test for whitespace like `1  234` which should throw a RuntimeException as well?

**Review by DanailMinchev — APPROVED** [review]
LGTM 👍 May be just few minor stuff which can be added later

File: src/test/java/uk/gov/pay/connector/model/FirstDigitsCardNumberTest.java (hunk 3 above)
Comment (simad) [reply src/test/java/uk/gov/pay/connector/model/FirstDigitsCardNumberTest.java:24]: I'll add that next :) thanks!

**Pull request merged**